src/utilities/filetypes/AlfalfaJSON.cpp

```diff
@@ -0,0 +1,223 @@
+
+        if (outFile) {
+          try {
+            outFile << root;
+            outFile.close();
+            return true;
+          } catch (...) {
+            LOG(Error, "Unable to write file to path '" << toString(m_JSONPath) << "'.");
+            return false;
+          }
+        }
+      }
+      LOG(Error, "Unable to write file to path '" << toString(m_JSONPath) << "', because parent directory "
+                                                  << "could not be created.");
+
+      return false;
+    }
+
+    Json::Value AlfalfaJSON_Impl::toJSON() const {
+      Json::Value root;
+      for (AlfalfaPoint point : m_points) {
+        // No guard here as the toJSON call will throw an exception if the id does not exist.
+        root[point.id().get()] = point.toJSON();
+      }
+      return root;
+    }
+  }  // namespace detail
+
+  AlfalfaJSON::AlfalfaJSON() : m_impl(std::shared_ptr<detail::AlfalfaJSON_Impl>(new detail::AlfalfaJSON_Impl())) {}
+
+  AlfalfaJSON::AlfalfaJSON(const std::string& s) : m_impl(std::shared_ptr<detail::AlfalfaJSON_Impl>(new detail::AlfalfaJSON_Impl(s))) {}
+
+  AlfalfaJSON::AlfalfaJSON(const openstudio::path& p) : m_impl(std::shared_ptr<detail::AlfalfaJSON_Impl>(new detail::AlfalfaJSON_Impl(p))) {}
+
+  boost::optional<AlfalfaPoint> AlfalfaJSON::exposeConstant(float value, const std::string& display_name) {
+    AlfalfaConstant component(value);
+    return exposeFromComponent(component, display_name);
+  }
+
+  boost::optional<AlfalfaPoint> AlfalfaJSON::exposeMeter(const std::string& meter_name, const std::string& display_name) {
+    AlfalfaMeter component(meter_name);
+    return exposeFromComponent(component, display_name);
+  }
+
+  boost::optional<AlfalfaPoint> AlfalfaJSON::exposeActuator(const std::string& component_name, const std::string& component_type,
+                                                            const std::string& control_type, const std::string& display_name) {
+    AlfalfaActuator component(component_name, component_type, control_type);
+    return exposeFromComponent(component, display_name);
+  }
+
+  boost::optional<AlfalfaPoint> AlfalfaJSON::exposeOutputVariable(const std::string& variable_key, const std::string& variable_name,
+                                                                  const std::string& display_name) {
+    AlfalfaOutputVariable component(variable_key, variable_name);
+    return exposeFromComponent(component, display_name);
+  }
+
+  boost::optional<AlfalfaPoint> AlfalfaJSON::exposeGlobalVariable(const std::string& variable_name, const std::string& display_name) {
+    AlfalfaGlobalVariable component(variable_name);
+    return exposeFromComponent(component, display_name);
+  }
+
+  void AlfalfaJSON::exposePoint(const AlfalfaPoint& point) {
+    m_impl->exposePoint(point);
+  }
+
+  boost::optional<AlfalfaPoint> AlfalfaJSON::exposeFromObject(const openstudio::IdfObject& idf_object, const std::string& display_name) {
```

Comment:
THe vast majority of the code I've read was "throwy", but here you go with LOG and return boost::none. That's fine, just noting.

Reply:
The general pattern was I log and return `boost::none` everywhere possible, but in ctors you can't do that (to my knowledge) to I throw a `std::runtime_error`.